controllers/cluster_controller.go

```diff
@@ -43,14 +45,18 @@ type ClusterReconciler struct {
 	validator               *vsphere.Validator
 	defaulter               *vsphere.Defaulter
 	tracker                 *remote.ClusterCacheTracker
-	buildProviderReconciler BuildProviderReconciler
+	buildProviderReconciler ProviderReconcilerBuilder
+}
+
+type ProviderClusterReconciler interface {
```

Comment:
I see what you are saying. However, these are not reconciling a datacenter config. They are reconciling a cluster, but following a different logic depending on the provider. If we had something like a `VSphereCluster` CRD, this would be its controller.

You make a good point, we don't have a resource to pair this reconcilers. But that extends to our API in general: we didn't design it with kubernetes controllers in mind (one controller per resource) but with the CLI and how the user interacts with config yaml file. But addressing that (which is doable) is a whole other topic.

I still think [Provider]ClusterReconciler is the right name for what these entities do. However, if we think we need to rename them, I would vote to continue that conversation outside of this PR. The goal here was to reorganize the code with a more sound package structure. We can add renaming refactors on top of it.